app/components/autocomplete/at_mention/at_mention.tsx

```diff
@@ -353,16 +380,37 @@ const AtMention = ({
     }, [cursorPosition]);
 
     useEffect(() => {
-        if (useGroupMentions) {
-            getGroupsForAutocomplete(serverUrl, channelId || '').then((res) => {
-                setGroups(res.length ? res : emptyGroupList);
-            }).catch(() => {
-                setGroups(emptyGroupList);
-            });
+        if (useGroupMentions && matchTerm && matchTerm !== '') {
+            // If the channel is constrained, we only show groups for that channel
+            if (isChannelConstrained && channelId) {
+                getFilteredChannelGroups(serverUrl, channelId, matchTerm).then((g) => {
+                    setGroups(g.length ? g : emptyGroupList);
+                }).catch(() => {
+                    setGroups(emptyGroupList);
+                });
+            }
+
+            // If there is no channel constraint, but a team constraint - only show groups for team
+            if (isTeamConstrained && !isChannelConstrained) {
+                getFilteredTeamGroups(serverUrl, teamId!, matchTerm).then((g) => {
+                    setGroups(g.length ? g : emptyGroupList);
+                }).catch(() => {
+                    setGroups(emptyGroupList);
+                });
+            }
+
+            // No constraints? Search all groups
+            if (!isTeamConstrained && !isChannelConstrained) {
+                getGroupsForAutocomplete(serverUrl, matchTerm || '').then((g) => {
+                    setGroups(g.length ? g : emptyGroupList);
+                }).catch(() => {
+                    setGroups(emptyGroupList);
+                });
+            }
         } else {
             setGroups(emptyGroupList);
         }
-    }, [channelId, useGroupMentions]);
+    }, [matchTerm, useGroupMentions]);
```

Comment:
Shouldn't we depend on `isChannelConstrained` and `isTeamConstrained` too? Probably not a big deal, since matchTerm will change often enough.

Reply:
Yup - plus channel / team constraints are set in the system console by admins so would be pretty rare to change often enough to cause issues here.